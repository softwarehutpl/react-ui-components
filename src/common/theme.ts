const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
