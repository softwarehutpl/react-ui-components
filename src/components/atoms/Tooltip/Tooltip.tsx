import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import {
  countTooltipArrowLeftPosition,
  countTooltipArrowTopPosition,
  countTooltipLeftPosition,
  countTooltipTopPosition,
} from './helpers';

interface ITooltipCommon {
  className?: string;
  color?: string;
  backgroundColor?: string;
  fontColor?: string;
  position?: string;
  opacity?: number;
}

interface ITooltipProps extends ITooltipCommon {
  targetElementId: string;
  tooltipText: string;
}

interface ITooltipContentStyleProps extends ITooltipCommon {
  isVisible: boolean;
  targetDomRect: { [key: string]: number };
  tooltipSize: { [key: string]: number };
  children: React.ReactNode;
}

const Tooltip = ({
  targetElementId,
  className,
  tooltipText,
  backgroundColor,
  color,
  fontColor,
  position,
  opacity,
}: ITooltipProps) => {
  const [isVisible, setIsVisible] = useState(false);
  const [targetDomRect, setTargetDomRect] = useState({});
  const [tooltipSize, setTooltipSize] = useState({});
  const tooltipRef = React.useRef<HTMLSpanElement>(null);

  const getTooltipSize = () => {
    if (tooltipRef && tooltipRef.current) {
      const { width, height } = tooltipRef.current.getBoundingClientRect();
      setTooltipSize({ width, height });
    }
  };

  const getTargetSize = (targetElement: any) => {
    const { top, left, right, bottom, width, height } = targetElement.getBoundingClientRect();
    setTargetDomRect({ top, left, right, bottom, width, height });
  };

  useEffect(() => {
    const targetElement = document.getElementById(targetElementId);
    if (targetElement) {
      targetElement.addEventListener('mouseenter', () => {
        setIsVisible(true);
      });
      targetElement.addEventListener('mouseleave', () => {
        setIsVisible(false);
      });
      getTooltipSize();
      getTargetSize(targetElement);
      window.addEventListener('resize', () => {
        getTooltipSize();
        getTargetSize(targetElement);
      });
    }
  }, [targetElementId]);

  return (
    <StyledTooltip
      className={className}
      isVisible={isVisible}
      backgroundColor={backgroundColor}
      color={color}
      fontColor={fontColor}
      targetDomRect={targetDomRect}
      position={position}
      ref={tooltipRef}
      tooltipSize={tooltipSize}
      opacity={opacity}
    >
      {tooltipText}
    </StyledTooltip>
  );
};

const defaultProps = {
  color: 'primary',
  position: 'top',
  opacity: 1,
};

Tooltip.defaultProps = defaultProps;

const TooltipContent = React.forwardRef<
  HTMLSpanElement,
  ITooltipCommon & ITooltipContentStyleProps
>(({ className, children }, ref) => {
  return (
    <span ref={ref} className={className}>
      {children}
    </span>
  );
});

const StyledTooltip = styled(TooltipContent)<ITooltipContentStyleProps>`
  ${({
    theme,
    color = defaultProps.color,
    isVisible,
    backgroundColor,
    fontColor,
    position = defaultProps.position,
    targetDomRect,
    tooltipSize,
    opacity = defaultProps.opacity,
  }) => {
    const tooltipBackground = backgroundColor || theme.colors[color].base;
    return {
      visibility: isVisible ? 'visible' : 'hidden',
      padding: '10px 20px',
      'border-radius': '4px',
      'background-color': tooltipBackground,
      color: fontColor || theme.colors[color].light,
      position: 'fixed',
      top: countTooltipTopPosition(position, tooltipSize, targetDomRect) || 'auto',
      left: countTooltipLeftPosition(position, tooltipSize, targetDomRect) || 'auto',
      'font-size': '0.8em',
      opacity,
      '&::after': {
        content: '""',
        display: 'block',
        width: 0,
        height: 0,
        'border-style': 'solid',
        'border-top-width': position === 'bottom' ? 0 : '8px',
        'border-top-color': position === 'top' ? tooltipBackground : 'transparent',
        'border-bottom-width': position === 'top' ? 0 : '8px',
        'border-bottom-color': position === 'bottom' ? tooltipBackground : 'transparent',
        'border-right-width': position === 'left' ? 0 : '8px',
        'border-right-color': position === 'right' ? tooltipBackground : 'transparent',
        'border-left-width': position === 'right' ? 0 : '8px',
        'border-left-color': position === 'left' ? tooltipBackground : 'transparent',
        position: 'absolute',
        left: countTooltipArrowLeftPosition(position, tooltipSize.width) || 'auto',
        top: countTooltipArrowTopPosition(position, tooltipSize.height) || 'auto',
      },
    };
  }}
`;

export default Tooltip;
