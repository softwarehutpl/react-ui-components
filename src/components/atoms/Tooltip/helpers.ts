const tooltipOffset = 15;
const arrowSize = 8;

export const countTooltipTopPosition = (
  position: string,
  tooltipSize: { [key: string]: number },
  targetDomRect: { [key: string]: number }
) => {
  if (position === 'top') {
    return `${targetDomRect.top - tooltipSize.height - tooltipOffset}px`;
  }
  if (position === 'bottom') {
    return `${targetDomRect.bottom + tooltipOffset}px`;
  }
  return `${targetDomRect.top + targetDomRect.height / 2 - tooltipSize.height / 2}px`;
};

export const countTooltipLeftPosition = (
  position: string,
  tooltipSize: { [key: string]: number },
  targetDomRect: { [key: string]: number }
) => {
  if (position === 'left') {
    return `${targetDomRect.left - tooltipSize.width - tooltipOffset}px`;
  }
  if (position === 'right') {
    return `${targetDomRect.right + tooltipOffset}px`;
  }
  return `${targetDomRect.left + targetDomRect.width / 2 - tooltipSize.width / 2}px`;
};

export const countTooltipArrowTopPosition = (position: string, tooltipHeight: number) => {
  if (position === 'top') {
    return `${tooltipHeight}px`;
  }
  if (position === 'bottom') {
    return `${-arrowSize}px`;
  }
  return `${tooltipHeight / 2 - arrowSize }px`;
};

export const countTooltipArrowLeftPosition = (position: string, tooltipWidth: number) => {
  if (position === 'left') {
    return `${tooltipWidth}px`;
  }
  if (position === 'right') {
    return `${-arrowSize}px`;
  }
  return `${tooltipWidth / 2 - arrowSize }px`;
};
