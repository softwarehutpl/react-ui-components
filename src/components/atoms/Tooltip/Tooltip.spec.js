import React from 'react';
import { create } from 'react-test-renderer';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { fireEvent, render } from '@testing-library/react';
import Tooltip from './Tooltip';
import theme from '../../../common/theme';

describe('Tooltip component', () => {
  it('should render correctly', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Tooltip tooltipText="tooltip text" targetElementId="some_id" />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render tooltip with given text', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Tooltip tooltipText="tooltip text" />
      </ThemeProvider>
    );
    const instance = component.root;
    const tooltip = instance.findByType('span');
    expect(tooltip.props.children).toBe('tooltip text');
  });

  it ('should render tooltip with given class', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Tooltip className="test-class" />
      </ThemeProvider>
    );
    const instance = component.root;
    const tooltip = instance.findByType(Tooltip);
    expect(tooltip.props.className).toEqual(expect.stringContaining('test-class'));
  })

  it('should show and hide tooltip on hover target element', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Tooltip targetElementId="target_el" tooltipText="text" />
        <span id="target_el">target span</span>
      </ThemeProvider>
    );
    const targetSpan = getByText('target span');
    const tooltip = getByText('text');

    expect(tooltip).toHaveStyle('visibility: hidden');
    fireEvent.mouseEnter(targetSpan);
    expect(tooltip).toHaveStyle('visibility: visible');
    fireEvent.mouseLeave(targetSpan);
    expect(tooltip).toHaveStyle('visibility: hidden');
  })

  it('should render tooltip with default style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Tooltip />
      </ThemeProvider>
    );
    const tooltip = component.toJSON();
    expect(tooltip).toHaveStyleRule('background-color', '#000000');
    expect(tooltip).toHaveStyleRule('color', '#ffffff');
    expect(tooltip).toHaveStyleRule('opacity', '1');
  });

  it('should render tooltip with custom style', () => {
    const component = create(
      <ThemeProvider theme={theme}>
        <Tooltip backgroundColor="red" fontColor="orange" opacity={0.5} />
      </ThemeProvider>
    );
    const tooltip = component.toJSON();
    expect(tooltip).toHaveStyleRule('background-color', 'red');
    expect(tooltip).toHaveStyleRule('color', 'orange');
    expect(tooltip).toHaveStyleRule('opacity', '0.5');
  });
});
