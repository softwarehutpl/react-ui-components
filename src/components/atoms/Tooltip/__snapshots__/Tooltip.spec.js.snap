// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component should render correctly 1`] = `
.c0 {
  visibility: hidden;
  padding: 10px 20px;
  border-radius: 4px;
  background-color: #000000;
  color: #ffffff;
  position: fixed;
  top: NaNpx;
  left: NaNpx;
  font-size: 0.8em;
  opacity: 1;
}

.c0::after {
  content: "";
  display: block;
  width: 0;
  height: 0;
  border-style: solid;
  border-top-width: 8px;
  border-top-color: #000000;
  border-bottom-width: 0;
  border-bottom-color: transparent;
  border-right-width: 8px;
  border-right-color: transparent;
  border-left-width: 8px;
  border-left-color: transparent;
  position: absolute;
  left: NaNpx;
  top: undefinedpx;
}

<span
  className="c0"
>
  tooltip text
</span>
`;
